File: examples/elixir/basic/test/basic_types_test.exs
defmodule BasicTypesTest do
  use ExUnit.Case, async: true

  # ── BasicItem struct ─────────────────────────────────────────────────────────

  describe "BasicItem struct" do
    test "can be created with all fields" do
      item = %Basic.BasicItem{name: "hello", count: 42}
      assert item.name == "hello"
      assert item.count == 42
    end

    test "count defaults to nil when only required fields are set" do
      item = %Basic.BasicItem{name: "hello"}
      assert item.count == nil
    end

    test "supports struct update syntax" do
      item0 = %Basic.BasicItem{name: "a", count: 1}
      item1 = %{item0 | count: 2}
      assert item1.name == "a"
      assert item1.count == 2
    end
  end

  # ── BasicStatus (string enum as tagged-union type) ───────────────────────────

  describe "BasicStatus enum" do
    test "known wire values map to atoms" do
      assert Basic.BasicStatus.from_string("ACTIVE") == :active
      assert Basic.BasicStatus.from_string("INACTIVE") == :inactive
      assert Basic.BasicStatus.from_string("PENDING") == :pending
    end

    test "unknown wire values round-trip via {:unknown, value}" do
      assert Basic.BasicStatus.from_string("FUTURE_STATUS") ==
               {:unknown, "FUTURE_STATUS"}

      assert Basic.BasicStatus.to_string({:unknown, "FUTURE_STATUS"}) ==
               "FUTURE_STATUS"
    end

    test "values/0 returns all known atoms" do
      assert Basic.BasicStatus.values() == [:active, :inactive, :pending]
    end
  end

  # ── BasicPriority (integer enum as tagged-union type) ────────────────────────

  describe "BasicPriority intEnum" do
    test "known integer values map to atoms" do
      assert Basic.BasicPriority.from_integer(1) == :low
      assert Basic.BasicPriority.from_integer(2) == :medium
      assert Basic.BasicPriority.from_integer(3) == :high
    end

    test "unknown integer values round-trip via {:unknown, value}" do
      assert Basic.BasicPriority.from_integer(99) == {:unknown, 99}
      assert Basic.BasicPriority.to_integer({:unknown, 99}) == 99
    end

    test "values/0 returns all known atoms" do
      assert Basic.BasicPriority.values() == [:low, :medium, :high]
    end
  end

  # ── BasicUnion (tagged union) ────────────────────────────────────────────────

  describe "BasicUnion tagged tuple" do
    test "text variant" do
      v = {:text, "hello"}
      assert {:text, str} = v
      assert str == "hello"
    end

    test "number variant" do
      v = {:number, 42}
      assert {:number, n} = v
      assert n == 42
    end

    test "flag variant" do
      v = {:flag, true}
      assert {:flag, b} = v
      assert b == true
    end

    test "unknown variant preserves wire tag" do
      v = {:unknown, "new_field"}
      assert {:unknown, tag} = v
      assert tag == "new_field"
    end
  end
end

File: examples/erlang/basic/test/basic_types_test.erl
-module(basic_types_test).

-include_lib("eunit/include/eunit.hrl").
-include("basic_types.hrl").

%% ── basic_not_found error record ─────────────────────────────────────────────

basic_not_found_record_test() ->
    Err = #basic_not_found{
        message = <<"missing">>,
        '__beam_error_kind' = client
    },
    ?assertEqual(<<"missing">>, Err#basic_not_found.message),
    ?assertEqual(client, Err#basic_not_found.'__beam_error_kind').

basic_not_found_default_kind_test() ->
    Err = #basic_not_found{message = <<"gone">>},
    ?assertEqual(client, Err#basic_not_found.'__beam_error_kind').

%% ── basic_item record ────────────────────────────────────────────────────────

basic_item_all_fields_test() ->
    Item = #basic_item{name = <<"hello">>, count = 42},
    ?assertEqual(<<"hello">>, Item#basic_item.name),
    ?assertEqual(42,          Item#basic_item.count).

basic_item_optional_count_test() ->
    Item = #basic_item{name = <<"hello">>},
    ?assertEqual(undefined, Item#basic_item.count).

basic_item_update_test() ->
    Item0 = #basic_item{name = <<"a">>, count = 1},
    Item1 = Item0#basic_item{count = 2},
    ?assertEqual(<<"a">>, Item1#basic_item.name),
    ?assertEqual(2,       Item1#basic_item.count).

%% ── basic_status (string enum as tagged-union type) ─────────────────────────
%% Erlang types are not enforced at runtime; the tests below verify that values
%% matching the type spec can be constructed and matched without error.

basic_status_known_values_test() ->
    Values = [active, inactive, pending],
    lists:foreach(fun(V) ->
        %% Pattern-match each known atom to confirm it is a valid Erlang term.
        ?assert(is_atom(V))
    end, Values).

basic_status_unknown_variant_test() ->
    Unknown = {unknown, <<"FUTURE_STATUS">>},
    {unknown, Wire} = Unknown,
    ?assertEqual(<<"FUTURE_STATUS">>, Wire).

%% ── basic_priority (integer enum as tagged-union type) ───────────────────────

basic_priority_known_values_test() ->
    Values = [low, medium, high],
    lists:foreach(fun(V) ->
        ?assert(is_atom(V))
    end, Values).

basic_priority_unknown_variant_test() ->
    Unknown = {unknown, 99},
    {unknown, Int} = Unknown,
    ?assertEqual(99, Int).

%% ── basic_union (tagged union) ───────────────────────────────────────────────

basic_union_text_variant_test() ->
    V = {text, <<"hello">>},
    {text, Str} = V,
    ?assertEqual(<<"hello">>, Str).

basic_union_number_variant_test() ->
    V = {number, 42},
    {number, N} = V,
    ?assertEqual(42, N).

basic_union_flag_variant_test() ->
    V = {flag, true},
    {flag, B} = V,
    ?assertEqual(true, B).

basic_union_unknown_variant_test() ->
    V = {unknown, <<"new_field">>},
    {unknown, Tag} = V,
    ?assertEqual(<<"new_field">>, Tag).
